checkpoint/runner.go

```diff
@@ -0,0 +1,154 @@
+package checkpoint
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+
+	"github.com/spf13/afero"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/accounts"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+)
+
+const (
+	SchemaVersion = "https://spacemesh.io/checkpoint.schema.json.1.0"
+
+	checkpointDir = "checkpoint"
+	schemaFile    = "schema.json"
+	dirPerm       = 0o700
+)
+
+type Runner struct {
+	fs     afero.Fs
+	db     *sql.Database
+	logger log.Log
+	dir    string
+}
+
+type Opt func(*Runner)
+
+func WithDataDir(dir string) Opt {
+	return func(r *Runner) {
+		r.dir = dir
+	}
+}
+
+func WithLogger(lg log.Log) Opt {
+	return func(r *Runner) {
+		r.logger = lg
+	}
+}
+
+func WithFilesystem(fs afero.Fs) Opt {
+	return func(r *Runner) {
+		r.fs = fs
+	}
+}
+
+func NewRunner(db *sql.Database, opts ...Opt) *Runner {
+	r := &Runner{
+		fs:     afero.NewOsFs(),
+		db:     db,
+		logger: log.NewNop(),
+	}
+	for _, opt := range opts {
+		opt(r)
+	}
+	return r
+}
+
+func (r *Runner) Generate(ctx context.Context, snapshotLayer types.LayerID, restoreLayer types.LayerID) ([]byte, error) {
+	checkpoint := Checkpoint{
+		Version: SchemaVersion,
+		Data: InnerData{
+			CheckpointId: fmt.Sprintf("snapshot-%d-restore-%d", snapshotLayer, restoreLayer),
+			Restore:      restoreLayer.Uint32(),
```

Comment:
it is orthogonal yes, just convenient for systest.

when we deliver this checkpoint data, we can change the restore layer to whatever we want. but since i intend to deliver the checkpoint as a file w. cmd line option (not rpc).

Reply:
it will have to be explained to everyone who will generate this thing, i would rather split this two operations, but don't have clear suggestion how exactly

Reply:
the node operator should get this file from the spacemesh url.
for ppl who wants to generate the checkpoint file to verify, they can specify this in the RPC request to create the exact same file.

but will try harder to decouple this in the future. i also think it's best to decouple these two settings.